src/Swarm/TUI/Controller.hs

```diff
@@ -102,238 +102,245 @@ pattern FKey :: Int -> BrickEvent n e
           mx = snd mouseLoc' + fst regionStart
           my = fst mouseLoc' + snd regionStart
-       in Just $ W.Coords (mx, my)
+       in pure . Just $ W.Coords (mx, my)
 
-setFocus :: AppState -> Name -> EventM Name (Next AppState)
-setFocus s name = continue $ s & uiState . uiFocusRing %~ focusSetCurrent name
+setFocus :: Name -> EventM Name AppState ()
+setFocus name = uiState . uiFocusRing %= focusSetCurrent name
 
 -- | Set the game to Running if it was auto paused
-maybeUnpause :: AppState -> EventM Name AppState
-maybeUnpause s
-  | s ^. gameState . runStatus == AutoPause = do
+maybeUnpause :: EventM Name AppState ()
+maybeUnpause = do
+  run <- use $ gameState . runStatus
+  when (run == AutoPause) $ do
     curTime <- liftIO $ getTime Monotonic
-    pure $ s & (gameState . runStatus .~ Running) . resetLastFrameTime curTime
-  | otherwise = pure s
+    resetLastFrameTime curTime
+    gameState . runStatus .= Running
  where
   -- When unpausing, it is critical to ensure the next frame doesn't
   -- catch up from the time spent in pause.
   -- TODO: manage unpause more safely to also cover
   -- the world event handler for the KChar 'p'.
-  resetLastFrameTime curTime = uiState . lastFrameTime .~ curTime
-
-toggleModal :: AppState -> ModalType -> EventM Name AppState
-toggleModal s mt = case s ^. uiState . uiModal of
-  Nothing -> pure $ s & (uiState . uiModal ?~ generateModal s mt) . ensurePause
-  Just _ -> maybeUnpause s <&> uiState . uiModal .~ Nothing
+  resetLastFrameTime curTime = uiState . lastFrameTime .= curTime
+
+toggleModal :: ModalType -> EventM Name AppState ()
+toggleModal mt = do
+  modal <- use $ uiState . uiModal
+  case modal of
+    Nothing -> do
+      newModal <- gets $ flip generateModal mt
+      ensurePause
+      uiState . uiModal ?= newModal
+    Just _ -> uiState . uiModal .= Nothing >> maybeUnpause
  where
   -- these modals do not pause the game
   runningModals = [RobotsModal]
   -- Set the game to AutoPause if needed
-  ensurePause
-    | s ^. gameState . paused || mt `elem` runningModals = id
-    | otherwise = gameState . runStatus .~ AutoPause
+  ensurePause = do
+    pause <- use $ gameState . paused
+    unless (pause || mt `elem` runningModals) $ do
+      gameState . runStatus .= AutoPause
 
-handleModalEvent :: AppState -> V.Event -> EventM Name (Next AppState)
-handleModalEvent s = \case
+handleModalEvent :: V.Event -> EventM Name AppState ()
+handleModalEvent = \case
   V.EvKey V.KEnter [] -> do
-    s' <- toggleModal s QuitModal
-    case s ^? uiState . uiModal . _Just . modalDialog . to dialogSelection of
-      Just (Just QuitButton) -> quitGame s'
-      Just (Just (NextButton scene)) -> startGame scene s'
-      _ -> continue s'
+    toggleModal QuitModal
+    modal <- use $ uiState . uiModal
+    case modal ^? _Just . modalDialog . to dialogSelection of
+      Just (Just QuitButton) -> quitGame
+      Just (Just (NextButton scene)) -> startGame scene
+      _ -> return ()
```

Comment:
Found it! The original code got the dialogue selection on the original state `s` not `s'`.

Reply:
Aha, right!